connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -266,31 +422,51 @@ public void close() {
         @Override
         public void onStartup(String connector) {
             state = AbstractStatus.State.RUNNING;
-            delegate.onStartup(connector);
+            synchronized (this) {
```

Comment:
The locking here is to ensure that, upon being cancelled, the very last status update that this connector makes is to set its state to `UNASSIGNED`.

There's a potential race if its status is scheduled to be updated to, e.g., `PAUSED` and the check for `cancelled` goes through because it isn't set to `true` yet, then the task gets cancelled on another thread and its status gets set to `UNASSIGNED`, then the original thread proceeds with execution and sets the status to `PAUSED`.

Reply:
I don't think locking buys you anything w/r/t to that. The producer#send in the status backing store is asynchronous. So what would describe above can happen anyways, regardless of whether you lock this object. 

Of course, if it wasn't asynchronous things would be much worse. A bottleneck would be created by the lock, waiting for the `send` to finish in every locked block, so that's not an option. Wdyt? 
That's what I see at the high level without spending to much time on it, but see if you can check this assumption and we can return to this question.

Reply:
AFAICT the underlying `KafkaBasedLog` used for the status store does provide ordering guarantees: https://github.com/apache/kafka/blob/1d438033f7f4d7cbdac34e1b57d97b5a647ae4de/connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java#L243-L247

Reply:
Thanks for checking the underlying implementation @C0urante . 

That takes us to my earlier concern about this operation potentially blocking for too long to be in a `synchronized` block. And the potential of blocking does not have to do with acknowledging that the record was written only. The producer call has a metadata update call too. 

Going over the uses of `KafkaBasedLog` in Connect, I didn't find an example where we have `KafkaBasedLog#send` running  in mutual exclusion. Contrary, similar concerns are probably the reason why we call `OffsetStorageWriter#doFlush` outside the synchronized block in `WorkerSourceTask`. 

I think we might be able to live with a rare race condition as the one you described, in order to avoid introducing unintended side-effects due to locking.